### PR TITLE
Override vimwiki's CTRL+Space to mark task as done when already started.

### PR DESCRIPTION
**Pull request by aaronfc**
Hi,
I have been working on this #175 for a couple of days now.

The change was pretty straightforward but I had some trouble with the tests. I found out some issues along the way and fixed them.

Here you have the details:
- Actual implementation of the <C-Space> override. Based on the same code used for <CR> override (added test).
- Fix test initialization for multi-syntax tests. Markup was being set _after_ the setup call and thus making the test fail if it was not reset by check_sanity first. Previously it was working because markup syntax was being set to "None" (as a string) and breaking TaskCache initialization **always** for the first time, which made check_sanity to restart the vim instance.
- Fix vimwiki initialization for tests. Init script is not being called by default, so you need to make it explicit.
- Fix travis build. You can see a full run here: https://travis-ci.org/github/aaronfc/taskwiki/builds/673640454.

Regarding the travis build, I decided to include the xvfb inside the docker-image itself so we do not need to rely on it running on travis. A side effect (which I find good) is that now tests run without opening an actual gvim screen and you do not need to run `xhost` command anymore.

A part from that you can see that there are three tests failing. I took a look at them and here you have what I saw:

FAILED tests/test_selected.py::TestInfoActionNotTriggeredByEnterOnLink::test_execute
This is not working and I do not know why. When pressing <CR> VimwikiFollowLink is being called, but it is not doing anything. I have always used vimwiki/taskwiki with Markdown support and this is an empty wikimedia link case. No idea what is happening, but as this is calling VimwikiFollowLink I am assuming it is related to Vimwiki or its configuration.

FAILED tests/test_selected.py::TestSelectAfterBufferSwitch::test_execute
This I think it is because `set filetype=vimwiki` is in fact returning a couple of empty lines. And it is expected to be silent by the test. I would suggest not checking if it is silent or not as it relies on changes on vimwiki itself.

FAILED tests/test_viewport.py::TestViewportsTaskGenerationEmptyFilter::test_execute[markdown]
This I tested and it is in fact not working. If I set an empty filter in markdown (on my local machine using default's master branch) it is not displaying the tasks. We need a fix for this. Shall we create a ticket?

Sorry for the long text, I have been working on this for a couple of days and many things raised. Do not hesitate to discuss any of the above!

### CODE REVIEW

File: ftplugin/vimwiki/taskwiki.vim [hunk 1]
```diff
@@ -103,6 +110,8 @@ if !exists('g:taskwiki_suppress_mappings')
                         let maplocalleader = '\t'
                 endif
         endif
+        " Override vimwiki's <C-Space>
+        nnoremap <silent><buffer> <C-Space> :TaskWikiDone<CR>
```
Comment (liskin) [inline ftplugin/vimwiki/taskwiki.vim:114]: Why this second mapping? Isn't the one on line 100 enough?

Comment (aaronfc) [reply ftplugin/vimwiki/taskwiki.vim]: That's a good question... I didn't understand very well this part at first but now re-reading it I think you are right. I just removed these lines and launched a new build[1]. The tests I added pass so probably I messed that up, thanks for pointing out!
Updating PR!

[1]: https://travis-ci.org/github/aaronfc/taskwiki/builds/699457085

**Comment by liskin** [comment]
Hey @aaronfc, unrelated but the issues on your fork are closed:

Would you mind cleaning up the branches in your fork? There's a branch which includes a complete history of a completely different and unrelated large open source project and this makes cloning your repo noticeably slower than necessary. See #269. Thanks!

**Comment by liskin** [comment]
> FAILED tests/test_selected.py::TestInfoActionNotTriggeredByEnterOnLink::test_execute
> This is not working and I do not know why. When pressing VimwikiFollowLink is being called, but it is not doing anything. I have always used vimwiki/taskwiki with Markdown support and this is an empty wikimedia link case. No idea what is happening, but as this is calling VimwikiFollowLink I am assuming it is related to Vimwiki or its configuration.

This test fails because it's creating a new wiki page and thus needs `g:vimwiki_list` to be setup correctly, but it's not because of quoting. A fix for this (and many other issues related tests) is in my [tests](https://github.com/tbabej/taskwiki/compare/master...liskin:tests) branch. I'll create a pull request later.

Thanks a lot for your fixes here, I took some inspiration from them.

**Comment by aaronfc** [comment]
> Hey @aaronfc, unrelated but the issues on your fork are closed:
> 
> Would you mind cleaning up the branches in your fork? There's a branch which includes a complete history of a completely different and unrelated large open source project and this makes cloning your repo noticeably slower than necessary. See #269. Thanks!
Done! I just removed the "naxisj_int_extraction" branch from my fork. Is that enough? I am not very experienced with forking and pull-requests here in Github so let me know if there is anything else missing.

Thank you for taking a look at this! :)

**Comment by liskin** [comment]
It's okay now, I can fetch from your fork without having to wait tens of seconds. Thanks! :-)

**Comment by aaronfc** [comment]
@liskin can you approve the merge? or we have to wait for @tbabej to approve?

**Comment by liskin** [comment]
I can't. I'm just another user who happens to have had some free time lately so I worked on taskwiki a lot, but I'm not affiliated with the project, so I can't merge anything here.

**Comment by aaronfc** [comment]
I do not understand why this is taking so long. And now I just saw your branch was merged... @tbabej ? con you comment on this?

**Comment by tbabej** [comment]
@aaronfc I merged liskin's branch as it was the most important out of the outstanding PRs - it actually fixed the tests (although it seems to have broken the coverage tracking, but that's a separate issue). This will help us to clear the open PR backlog, because we will be able to ensure we're not breaking existing functionality.

Currently merging of this branch is blocked - it needs to be rebased on top of the current master. Could you look into that? Otherwise I have no problem with this PR and it could be merged.

**Comment by liskin** [comment]
@tbabej Can you elaborate about the broken coverage tracking? https://coveralls.io/github/tbabej/taskwiki looks okay-ish to me. I'll be happy to look into that. (I've been fighting with coveralls and github actions for the last two days, actually.)

**Comment by tbabej** [comment]
@liskin Sure thing! I didn't have too much time to look into this, but couple of quick observations:
* You can see actually on the page you link that the coverage has decreased by roughly 58% (from 98% to 40% currently).
* If you look at which lines are actually covered, it's mostly lines that would be executed upon module's import (such as class / method definitions), but not actual lines that need methods themselves to be executed (the bodies of the methods are rarely marked as covered)

So I would guess the current coverage that is detected stems from tests importing taskwiki modules, and not from the individual vim sessions invoked by vimrunner? It's possible there is some bug in the collection of the individual coverage files, note that each test generates a `/tmp/taskwiki-coverage/.coverage.<PID>` file which need to be later combined in order to produce the final report.

@aaronfc Sorry for the off-topic!

**Comment by aaronfc** [comment]
> @aaronfc I merged liskin's branch as it was the most important out of the outstanding PRs - it actually fixed the tests (although it seems to have broken the coverage tracking, but that's a separate issue). This will help us to clear the open PR backlog, because we will be able to ensure we're not breaking existing functionality.
> 
> Currently merging of this branch is blocked - it needs to be rebased on top of the current master. Could you look into that? Otherwise I have no problem with this PR and it could be merged.

Will do the rebase when I find some time. Thanks!

**Comment by liskin** [comment]
Hey @aaronfc, I found an alternative approach to your solution:

Setting `let g:vimwiki_listsyms = ' WSX'` makes vimwiki's Ctrl+Space handle started/waiting tasks.
There's a problem, however: vimwiki automatically updates parent task's checkbox, which is likely not what we want with taskwiki. But it doesn't only do that on Ctrl-Space, it does it also on `o` and `O`, so I reported https://github.com/vimwiki/vimwiki/issues/954 and will try to submit a pull request there. Then we can set these variables from taskwiki and we won't need to remap this.

What do you think?

**Comment by liskin** [comment]
I forgot to post an update here: https://github.com/vimwiki/vimwiki/pull/1041 was merged a while ago so this now works well with these vimwiki options:

```
let g:vimwiki_list[0].listsym_rejected = 'D'
let g:vimwiki_list[0].listsyms = ' WSX'
let g:vimwiki_list[0].listsyms_propagate = 0
```

I'm not entirely sure why I didn't submit a PR here. Possibly I thought taskwiki shouldn't be overriding these settings? I don't remember. :-(

**Pull request closed without merging**